Enable `in_dims` for vmap frontend api

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40717 Enable `in_dims` for vmap frontend api**

`in_dims` specifies which dimension of the input tensors should be
vmapped over. One can also specify `None` as an `in_dim` for a particular
input to indicate that we do not map over said input.

We implement `in_dims` by creating a BatchedTensor with BatchDim equal
to said `in_dim`. Most of this PR is error checking. `in_dims` must
satisfy the following:
- `in_dim` can be either an int or a Tuple[Optional[int]]. If it is an
int, we use it to mean the `in_dim` for every input.
- If `in_dims` is not-None at some index `idx`, then the input at index
`idx` MUST be a tensor (vmap can only map over tensors).

jax supports something more generalized: their `in_dims` can match the
structure of the `inputs` to the function (i.e., it is a nested python
data structure matching the data structure of `inputs` specifying where
in `inputs` the Tensors to be mapped are and what their map dims should
be). We don't have the infrastruture yet so we only support `int` or a
flat tuple for `in_dims`.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22397914](https://our.internmc.facebook.com/intern/diff/D22397914)

zou3519 committed:
Enable `in_dims` for vmap frontend api

`in_dims` specifies which dimension of the input tensors should be
vmapped over. One can also specify `None` as an `in_dim` for a particular
input to indicate that we do not map over said input.

We implement `in_dims` by creating a BatchedTensor with BatchDim equal
to said `in_dim`. Most of this PR is error checking. `in_dims` must
satisfy the following:
- `in_dim` can be either an int or a Tuple[Optional[int]]. If it is an
int, we use it to mean the `in_dim` for every input.
- If `in_dims` is not-None at some index `idx`, then the input at index
`idx` MUST be a tensor (vmap can only map over tensors).

jax supports something more generalized: their `in_dims` can match the
structure of the `inputs` to the function (i.e., it is a nested python
data structure matching the data structure of `inputs` specifying where
in `inputs` the Tensors to be mapped are and what their map dims should
be). We don't have the infrastruture yet so we only support `int` or a
flat tuple for `in_dims`.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Enable `in_dims` for vmap frontend api"


`in_dims` specifies which dimension of the input tensors should be
vmapped over. One can also specify `None` as an `in_dim` for a particular
input to indicate that we do not map over said input.

We implement `in_dims` by creating a BatchedTensor with BatchDim equal
to said `in_dim`. Most of this PR is error checking. `in_dims` must
satisfy the following:
- `in_dim` can be either an int or a Tuple[Optional[int]]. If it is an
int, we use it to mean the `in_dim` for every input.
- If `in_dims` is not-None at some index `idx`, then the input at index
`idx` MUST be a tensor (vmap can only map over tensors).

jax supports something more generalized: their `in_dims` can match the
structure of the `inputs` to the function (i.e., it is a nested python
data structure matching the data structure of `inputs` specifying where
in `inputs` the Tensors to be mapped are and what their map dims should
be). We don't have the infrastruture yet so we only support `int` or a
flat tuple for `in_dims`.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

ezyang commented on torch/_vmap_internals.py:
```diff
@@ -62,6 +108,30 @@ def _as_tuple(value, num_elements, error_message_lambda):
+            fn=fn_name, in_dims=in_dims, num_inputs=len(args)))
+
+    if len(args) == 0:
+        raise ValueError(NO_INPUTS.format(fn=fn_name))
```
Is there any reason to block the zero length args case, besides "then vmap doesn't do anything"? I'm thinking of how people have found it useful to do zero-size batches; it may be harmless to have zero length args (unless it is not?)

zou3519 replied:
The only two reasons I have are: (1) "then vmap doesn't do anything" and (2) "jax doesn't allow it". I agree that it seems harmless to have zero-length arguments.

It's not too hard to modify this to work so I'll add this as a follow-up for later (and think more about if it is actually harmless or not)

ezyang commented on torch/_vmap_internals.py:
```diff
@@ -34,19 +34,65 @@
+EXPECTED_IN_DIMS_TO_BE_INT_OR_TUPLE = (
+    'vmap({fn}, in_dims={in_dims}, ...): expected `in_dims` to be int or tuple, '
+    'got: {actual_type}.'
+)
```
No action needed on this comment: I personally prefer having messages inline at their use sites, if they're only used once. Makes it easier to see what the error message is and ensure that the format string is up to date :) (also, you can't use f-strings in this style!)

zou3519 replied:
Did we drop support for Python < 3.6? (I know we dropped support for Python 2, but I didn't realize the < 3.6 part)

I agree with your comment, reading 66 lines of error messages at the top of the file and away from the callsites makes me sad. Will fix in a follow-up.

ezyang commented on torch/_vmap_internals.py:
```diff
@@ -34,19 +34,65 @@
 
+# Check compatibility of `in_dims` and `args`. More specifically, checks the following:
+# Wherever an in_dim is not None, then the corresponding index in args must be
+# a Tensor. Furthermore, tensor must have the `in_dim` (0 <= in_dim < tensor.dim())
```
Type signature on this function (and the others) would be very helpful!

zou3519 replied:
RIP Python 2!!!

zou3519 replied:
I added type hints to all of the functions :D. We'll have to relax some of these in the future when we support accepting arbitrary nested python data structures, but these do make the code easier to read now.

ezyang commented on torch/_vmap_internals.py:
```diff
@@ -34,19 +34,65 @@
+# Wherever an in_dim is not None, then the corresponding index in args must be
+# a Tensor. Furthermore, tensor must have the `in_dim` (0 <= in_dim < tensor.dim())
+def _check_args_can_be_mapped_with_in_dims(in_dims_as_tuple, args, fn_name, in_dims):
+    for idx, (in_dim, arg) in enumerate(zip(in_dims_as_tuple, args)):
```
If you extend this to work on arbitrary Python collections as opposed to just tuples, zipping here isn't going to work anymore, right? Would we expect in-dims to also have the same "shape" as args, in this case?

zou3519 replied:
That's correct. Extending this to work on arbitrary Python collections would make it so that we need new error validation code here. Furthermore, we'd expect in_dims to have the same "shape" as args.

zou3519 committed:
Update on "Enable `in_dims` for vmap frontend api"


`in_dims` specifies which dimension of the input tensors should be
vmapped over. One can also specify `None` as an `in_dim` for a particular
input to indicate that we do not map over said input.

We implement `in_dims` by creating a BatchedTensor with BatchDim equal
to said `in_dim`. Most of this PR is error checking. `in_dims` must
satisfy the following:
- `in_dim` can be either an int or a Tuple[Optional[int]]. If it is an
int, we use it to mean the `in_dim` for every input.
- If `in_dims` is not-None at some index `idx`, then the input at index
`idx` MUST be a tensor (vmap can only map over tensors).

jax supports something more generalized: their `in_dims` can match the
structure of the `inputs` to the function (i.e., it is a nested python
data structure matching the data structure of `inputs` specifying where
in `inputs` the Tensors to be mapped are and what their map dims should
be). We don't have the infrastruture yet so we only support `int` or a
flat tuple for `in_dims`.

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]
